### DISCUSSION TITLE
Why login with root on the servers?

### DISCUSSION THREAD

**Discussion by gefloh** [kube-hetzner/terraform-hcloud-kube-hetzner#678] [Q&A] [answered]
Hi,

just recently stumbled over this great project, amazing work you guys are doing, thanks a lot! 

Just one question which came into my mind: In all server hardening guides I always read that you should disable root login on a server. Here all my servers are provisioned with public IP addresses and enabled root login. Isn't this a security risk? Or are there technical reasons why the root login is necessary e.g. for Terraform?

Cheers,
Gerhard

**Comment by mysticaltech** [comment, marked as answer]
@gefloh We just chose to keep the default as given by OpenSUSE MicroOS, as it's easier for configuration. Here's how it's hardened:

openSUSE MicroOS, based on Tumbleweed, is designed to be an immutable, transactional operating system optimized for container and edge computing. While it's true that it comes with a root user by default, the security hardening measures in place help mitigate potential risks.

Some of these security features include:

- Read-only root filesystem: The root filesystem is set to read-only by default, which prevents unauthorized changes to critical system files.
- Atomic updates: MicroOS uses transactional updates that are atomic, meaning they are either applied completely or not at all. This ensures that the system is always in a consistent and stable state.
- Rollback capabilities: If an update causes issues, the system can easily be rolled back to a previous working state, minimizing the impact of any security vulnerabilities introduced by a new update.
- Containerization: MicroOS is designed to run applications in containers, which provides an additional layer of isolation from the host system.
- Regular updates: As MicroOS is based on Tumbleweed, it receives frequent updates, ensuring that the latest security patches are applied promptly.
- Hardened kernel: The kernel used in MicroOS comes with several security features enabled, such as Address Space Layout Randomization (ASLR), which makes it more resistant to potential attacks.
- SELinux enforcing: MicroOS can come with SELinux (Security-Enhanced Linux) in enforcing mode, which means that the system will actively prevent unauthorized actions based on its security policy. This provides an additional layer of protection against potential security vulnerabilities, as it restricts processes and users to the minimum set of privileges required for their tasks.

While having a root user does introduce some level of risk, the security measures built into openSUSE MicroOS help minimize the potential impact.

**Reply by mysticaltech** [reply]
@gefloh Also, and that was the main reason, even though k3s can run as a non-root user, but there are some limitations and additional configurations required. By default, k3s is designed to run as a root user since it requires access to specific resources and capabilities that are typically restricted to the root user.

All in all, it made sense for us to stay with the default that is both secure and optimized for Kubernetes.

**Reply by gefloh** [reply]
Thanks for explaining @mysticaltech, I see. But there shouldn't be a problem closing the SSH ports of all nodes with the firewall manually as long as not running Terraform updates on the infrastructure right? Since updates appear to be applied automatically, I see not really a point in having an SSH port open (apart from Terraform).

**Reply by mysticaltech** [reply]
@gefloh Good question! MicroOS is has transactional-udpates, meaning the update is done to a snapshot, and it reboots to that snapshot, if all is good, great, if not and something breaks, it reverts to the old snapshot. You if you have your cluster in HA (3 Control-planes, >= 2 Agents) it's pretty safe!

Now, you can turn the ugprades off, you have variables for that in your kube.tf, and also manually, see the readme.

Also, for the firewall, definitely, after install the SSH port is not needed, so you can close that off, but my advice never touch things manually. Do it via the kube.tf, extra_firewall_rules variable, POST first apply, see https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/436#issuecomment-1379241029.

If you want to open it again, you just remove that blocking rule, and apply again, it will open up again!

**Reply by gefloh** [reply]
Ah thanks for the hint with the firewall, will do it that way indeed. 

Yes, I went HA mainly because of the updates (and because it's cooler ;-), although obviously it costs more. But I think it's worth it.

**Reply by gefloh** [reply]
So, closing SSH works as expected when applying kube.tf. What does not work is, removing that block in the kube.tf and apply again. It runs into a

> Error: unable to open remote client: couldn't establish a connection to the remote server: dial tcp x.x.x.x:xxxx: connect: connection timed out

then. It seems the only way is removing/changing that rule manually in the Hetzner Console (and then maybe apply again to make things consistent). That was probably also what you meant right?

**Reply by mysticaltech** [reply]
@gefloh I did not mean to open it up again manually, but you will have to do that if you first close down, touch anything other than the firewall in the kube.tf, and apply again. 

So ideally, once the firewall is closed, the kube.tf should be a no-go zone, unless you want to open it up again, then you apply, then you can start playing around with you kube.tf again. Is that clear enough?

**Reply by gefloh** [reply]
I am not completely sure @mysticaltech . It didn't work for me opening the firewall again with just applying the kube.tf and having changed there only the firewall and nothing else. That would certainly be my preferred way of handling it. So currently, if you block the configured SSH port with applying kube.tf you cannot apply the kube.tf again until you have manually adjusted the firewall (in the Hetzner Console) to accept SSH connections on that port again. Maybe it's a bug? Can you verify this behavior?

**Reply by mysticaltech** [reply]
If I missed something, please copy paste the complete output of the terraform apply where the error occurs, so that I can see the plan and what caused a component to want to run a remote-provisioner.

**Reply by mysticaltech** [reply]
@gefloh I will verify on my end, no problem!

**Reply by mysticaltech** [reply]
@gefloh You are right, once closed it need to be opened manually, and then you can apply and it will work. Kinda not ideal but it's the best we have for now haha.
